File: src/catch-order.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { initial, step, simulationStep } from "./game.js";
import { startFlight } from "./ball.js";
import { velocityUnit } from "./attributes.js";
const unit = 22.4 / 576;
function setup(distance, velocity) {
  const s = initial();
  s.players.forEach((p) => (p.stun = 100));
  Object.assign(s.players[7], { x: 0, z: 0, stun: 0, fx: 1, fz: 0 });
  Object.assign(s.ball, {
    x: distance * unit,
    z: 0,
    vx: velocity * velocityUnit,
    vz: 0,
    owner: -1,
    lastTouch: 16,
    speedTimer: 100,
  });
  startFlight(s.ball, false);
  return s;
}
test("a ball entering catch range during movement waits until the next tick", () => {
  const s = setup(17, -8);
  step(s, simulationStep, {}, [true, true]);
  assert.equal(s.ball.owner, -1);
  assert.ok(s.ball.x < 16 * unit);
  step(s, simulationStep, {}, [true, true]);
  assert.equal(s.ball.owner, 7);
});
test("a ball already in range is caught before it can move away", () => {
  const s = setup(15, 8);
  step(s, simulationStep, {}, [true, true]);
  assert.equal(s.ball.owner, 7);
  assert.equal(s.ball.vx, 0);
});
test("simulation uses team-two-first catch order at equal roster index", () => {
  const s = initial();
  s.players.forEach((p) => (p.stun = 100));
  for (const i of [7, 16]) Object.assign(s.players[i], { x: 0, z: 0, stun: 0 });
  Object.assign(s.ball, { x: 0, z: 0, h: 0.75 });
  step(s, simulationStep, {}, [true, true]);
  assert.equal(s.ball.owner, 16);
});

File: server/catch_order_test.go
package main

import "testing"

func TestCatchBeforeMovement(t *testing.T) {
	for _, c := range []struct {
		distance, velocity float64
		owner              int
	}{{17, -8, -1}, {15, 8, 7}} {
		s := initial()
		for i := range s.Players {
			s.Players[i].Stun = 100
		}
		p := &s.Players[7]
		p.X, p.Z, p.Stun = 0, 0, 0
		p.FX = 1
		p.FZ = 0
		s.Ball = Ball{X: c.distance * (22.4 / 576), VX: c.velocity * velocityUnit, Owner: -1, LastTouch: 16, SpeedTimer: 100}
		startFlight(&s.Ball, false)
		s.step(simulationStep, [2]Input{})
		if s.Ball.Owner != c.owner {
			t.Fatalf("distance %v owner %d", c.distance, s.Ball.Owner)
		}
		if c.owner < 0 {
			s.step(simulationStep, [2]Input{})
			if s.Ball.Owner != 7 {
				t.Fatal("missing next-tick catch")
			}
		}
	}
}
func TestInterleavedCatchOrder(t *testing.T) {
	s := initial()
	for i := range s.Players {
		s.Players[i].Stun = 100
	}
	for _, i := range []int{7, 16} {
		s.Players[i].X = 0
		s.Players[i].Z = 0
		s.Players[i].Stun = 0
	}
	s.Ball = Ball{H: .75, Owner: -1, LastTouch: -1}
	s.step(simulationStep, [2]Input{})
	if s.Ball.Owner != 16 {
		t.Fatal(s.Ball.Owner)
	}
}
